platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassloadingCache.kt

```diff
@@ -33,10 +33,10 @@ class KotlinScriptClassloadingCache @Inject constructor(
 ) {
 
     private
-    val cache: CrossBuildInMemoryCache<ProgramId, CompiledScript> = cacheFactory.newCache()
+    val cache: CrossBuildInMemoryCache<ProgramId, CompiledScript> = cacheFactory.newCache { reused -> reused.onReuse() }
 
     fun get(key: ProgramId): CompiledScript? =
-        cache.getIfPresent(key)?.also { it.onReuse() }
+        cache.getIfPresent(key)
```

Comment:
ℹ️ This is the main change, the `onReuse()` call will only happen when the script is reused again in the next build session